### PR TITLE
feature: Dataset metadata writing

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Updates `writer.py` to handle the new Dataset metadata! (This does NOT touch the collection files yet.)

This is the second step towards #36, and is a validation for some of the approaches I'm using.

*Estimated size: medium, 20-30 minutes. (Functional code is ~90 lines, the rest is unit testing!)*

## Changes
- `ColorizerDatasetWriter` will now (correctly) load in and update metadata from existing manifests.
  - Some fields (like the revision number, time last modified, and writer version) will be updated automatically. 
  - Some fields (like the name and time of creation) will be updated if missing.
- Added unit tests for behavior and edge cases.

## Keyfiles (start with these!)
- `utils.py` -> convenience method to merge dictionary objects so metadata can be easily overwritten
- `writer.py` -> writer automatically loads in and updates metadata fields

### CODE REVIEW

File: colorizer_data/writer.py [hunk 1]
```diff
@@ -449,9 +489,9 @@ def validate_dataset(
         Logs warnings to the console if any expected files are missing.
         """
         if self.manifest["times"] is None:
-            logging.warn("No times JSON information provided!")
+            logging.warning("No times JSON information provided!")
         if not os.path.isfile(self.outpath + "/" + self.manifest["times"]):
-            logging.warn(
+            logging.warning(
```
Comment (ShrimpCryptid) [inline colorizer_data/writer.py:494]: `logging.warn` is deprecated

File: colorizer_data/utils.py [hunk 2]
```diff
@@ -522,3 +522,22 @@ def cast_feature_to_info_type(
...
+    """Recursively merges key-value pairs of `b` into `a`, ignoring keys with `None` values."""
+    # This is basically a replacement for `{...a, ...b}` in JavaScript
+    if b == None or a == None:
+        return a
+    # Make shallow copy of a
+    a = {**a}
+
+    for key, value in b.items():
+        if isinstance(value, dict):
+            a[key] = merge_dictionaries(a[key], value)
```
Comment (rugeli) [inline colorizer_data/utils.py:540]: Would it be safer to use `a[key] = merge_dictionaries(a.get(key), value)`to avoid a KeyError in this case?

File: colorizer_data/writer.py [hunk 3]
```diff
@@ -381,18 +394,50 @@ def write_manifest(
...
                 "ColorizerDatasetWriter: The argument `num_frames` on `write_manifest` is deprecated and will be removed in the future! Please call `set_frame_paths(generate_frame_paths(num_frames))` instead."
             )
             self.set_frame_paths(generate_frame_paths(num_frames))
 
-        # Add the metadata
-        if metadata:
-            self.manifest["metadata"] = metadata.to_json()
+        # Automatically update metadata fields. This behavior can be overwritten using
```
Comment (rugeli) [inline colorizer_data/writer.py:413]: It's nice to have auto updates for these fields!

File: colorizer_data/writer.py [hunk 4]
```diff
@@ -381,18 +394,50 @@ def write_manifest(
...
+        self.metadata.last_modified = current_time
+        self.metadata._writer_version = CURRENT_VERSION
+
+        # Use default dataset name from writer constructor if no name was loaded
+        # (will be overridden by `metadata.name` argument if provided)
+        if self.metadata.name == None:
+            self.metadata.name = self.default_dataset_name
+
+        # Optionally merge new metadata with old
+        if metadata != None:
```
Comment (rugeli) [inline colorizer_data/writer.py:435]: We may want to use `is None` and `is not None` instead of `==` and `!=`, unless you have a specific reason for using the latter.

Comment (ShrimpCryptid) [reply colorizer_data/writer.py]: Fixed, thank you!

**Review by rugeli — APPROVED** [review]
LGTM!

**Review by blairlyons — APPROVED** [review]
